File: day1_part1_2016.hs

import Data.List

type Position = (Int, Int)
data Direction = North | East | South | West deriving (Enum)

turn :: Direction -> Char -> Direction
turn dir 'R' = toEnum $ (fromEnum dir + 1) `mod` 4
turn dir 'L' = toEnum $ (fromEnum dir - 1) `mod` 4

move :: Position -> Direction -> Int -> Position
move (x, y) North dist = (x, y + dist)
move (x, y) East dist = (x + dist, y)
move (x, y) South dist = (x, y - dist)
move (x, y) West dist = (x - dist, y)

parseInstruction :: String -> (Char, Int)
parseInstruction (dir:dist) = (dir, read dist)

applyInstruction :: (Position, Direction) -> (Char, Int) -> (Position, Direction)
applyInstruction ((x, y), dir) (turnDir, dist) = (move (x, y) newDir dist, newDir)
    where newDir = turn dir turnDir

main :: IO ()
main = do
    input <- readFile "input.txt"
    let instructions = map parseInstruction $ words $ map (\c -> if c == ',' then ' ' else c) input
        finalPos = foldl applyInstruction ((0, 0), North) instructions
        distance = abs (fst (fst finalPos)) + abs (snd (fst finalPos))
    print distance

File: objectivec/day1_part1_2016.m
#import <Foundation/Foundation.h>

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        NSString *input = [NSString stringWithContentsOfFile:@"input.txt" encoding:NSUTF8StringEncoding error:nil];
        NSArray *instructions = [input componentsSeparatedByString:@", "];
        
        int x = 0;
        int y = 0;
        int direction = 0; // 0: North, 1: East, 2: South, 3: West
        
        for (NSString *instruction in instructions) {
            unichar turn = [instruction characterAtIndex:0];
            int distance = [[instruction substringFromIndex:1] intValue];
            
            if (turn == 'R') {
                direction = (direction + 1) % 4;
            } else {
                direction = (direction + 3) % 4;
            }
            
            if (direction == 0) {
                y += distance;
            } else if (direction == 1) {
                x += distance;
            } else if (direction == 2) {
                y -= distance;
            } else {
                x -= distance;
            }
        }
        
        int distance = abs(x) + abs(y);
        printf("%d\n", distance);
    }
    return 0;
}